api/v1/api.go

```diff
@@ -0,0 +1,212 @@
+	errorCanceled    errorType = "canceled"
+	errorExec        errorType = "execution"
+	errorBadData     errorType = "bad_data"
+	errorInternal    errorType = "internal"
+	errorUnavailable errorType = "unavailable"
+	errorNotFound    errorType = "not_found"
+)
+
+type apiError struct {
+	typ errorType
+	err error
+}
+
+func (e *apiError) Error() string {
+	return fmt.Sprintf("%s: %s", e.typ, e.err)
+}
+
+var corsHeaders = map[string]string{
+	"Access-Control-Allow-Headers":  "Accept, Authorization, Content-Type, Origin",
+	"Access-Control-Allow-Methods":  "GET, POST, DELETE, OPTIONS",
+	"Access-Control-Allow-Origin":   "*",
+	"Access-Control-Expose-Headers": "Date",
+	"Cache-Control":                 "no-cache, no-store, must-revalidate",
+}
+
+// Enables cross-site script calls.
+func setCORS(w http.ResponseWriter) {
+	for h, v := range corsHeaders {
+		w.Header().Set(h, v)
+	}
+}
+
+// API provides registration of handlers for API routes.
+type API struct {
+	logger      log.Logger
+	MetricStore storage.MetricStore
+	Flags       map[string]string
+	StartTime   time.Time
+	BuildInfo   map[string]string
+}
+
+// New returns a new API.
+// logger can be nil and if log.logger is set to nil no logging is possible
+// MetricStore cannot be left empty
+// Flags is information of flags provide at initialization of pushgateway
+// BuildInfo consist of different information which are set at build time
+// StartTime is time of initialization of APIs
+func New(
+	l log.Logger,
+	ms storage.MetricStore,
+	flags map[string]string,
+	buildInfo map[string]string,
+) *API {
+	if l == nil {
+		l = log.NewNopLogger()
+	}
+
+	return &API{
+		StartTime:   time.Now(),
+		logger:      l,
+		MetricStore: ms,
+		Flags:       flags,
+		BuildInfo:   buildInfo,
+	}
+}
+
+// Register registers the API handlers under their correct routes
+// in the given router.
+func (api *API) Register(r *route.Router) {
+	wrap := func(f http.HandlerFunc) http.HandlerFunc {
+		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			setCORS(w)
+			f(w, r)
+		})
+	}
+
+	r.Options("/*path", wrap(func(w http.ResponseWriter, r *http.Request) {}))
+
+	r.Get("/status", wrap(api.status))
+	r.Get("/metrics", wrap(api.metrics))
+}
+
+type metrics struct {
+	Timestamp time.Time     `json:"time_stamp"`
+	Type      string        `json:"type"`
+	Help      string        `json:"help"`
+	Metrics   []*dto.Metric `json:"metrics"`
+}
+
+func (api *API) metrics(w http.ResponseWriter, r *http.Request) {
+	familyMaps := api.MetricStore.GetMetricFamiliesMap()
+	res := make([]interface{}, 0)
+	for _, v := range familyMaps {
+		metricResponse := make(map[string]interface{})
+		metricResponse["label"] = v.Labels
+		metricResponse["last_success"] = v.LastPushSuccess()
+		for name, metricValues := range v.Metrics {
+			uniqueMetrics := metrics{
+				Type:      metricValues.GobbableMetricFamily.Type.String(),
+				Help:      *metricValues.GobbableMetricFamily.Help,
```

Comment:
If no help string was provided, `metricValues.GobbableMetricFamily.Help` is a nil pointer, which will crash the unmarshaling. You have to guard against that here. For example:

```Go
Help:      (*dto.MetricFamily)(metricValues.GobbableMetricFamily).GetHelp(),
```